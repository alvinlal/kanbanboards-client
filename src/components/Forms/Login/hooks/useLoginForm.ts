import { yupResolver } from '@hookform/resolvers/yup';
import { AxiosError } from 'axios';
import { useState, useCallback } from 'react';
import { useForm, SubmitHandler, Control, FieldErrorsImpl } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { useAuth } from '../../../../hooks/useAuth/useAuth';
import { useErrorHandlers } from '../../../../hooks/useErrorHandlers/useErrorHandlers';
import { UnauthorizedException } from '../../../../types/UnauthorizedException';

import { loginSchema } from '../validators/loginSchema';

interface UseLoginForm {
  loading: boolean;
  isValid: boolean;
  errors: Partial<FieldErrorsImpl<LoginFormInputs>>;
  control: Control<LoginFormInputs, unknown>;
  handleContinueWithGoogle: () => void;
  submitForm: (e: React.FormEvent) => void;
}

export type LoginFormInputs = {
  email: string;
  password: string;
};

export const useLoginForm = (): UseLoginForm => {
  const {
    control,
    handleSubmit,
    setError,
    formState: { errors, isValid },
  } = useForm<LoginFormInputs>({
    mode: 'onChange',
    resolver: yupResolver(loginSchema),
  });
  const [loading, setLoading] = useState(false);
  const { logIn } = useAuth();
  const { handleRequestError } = useErrorHandlers();
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<LoginFormInputs> = useCallback(
    async ({ email, password }) => {
      try {
        setLoading(true);
        await logIn({ email, password });
        navigate('/');
      } catch (err) {
        const requestErrors = handleRequestError(err as AxiosError) as UnauthorizedException;

        if (requestErrors) {
          setError('password', {
            message: requestErrors.message,
          });
        }
      } finally {
        setLoading(false);
      }
    },
    [logIn, navigate, handleRequestError, setError]
  );

  const handleContinueWithGoogle = useCallback(() => {
    window.location.href = `${import.meta.env.VITE_API_ENDPOINT}/auth/google`;
  }, []);

  const submitForm = (e: React.FormEvent) => {
    e.preventDefault();
    handleSubmit(onSubmit)();
  };

  return {
    control,
    errors,
    isValid,
    loading,
    submitForm,
    handleContinueWithGoogle,
  };
};
