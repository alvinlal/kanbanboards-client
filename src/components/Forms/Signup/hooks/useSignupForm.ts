import { yupResolver } from '@hookform/resolvers/yup';
import { AxiosError } from 'axios';
import React, { useState, useCallback } from 'react';
import { useForm, SubmitHandler, FieldErrorsImpl, Control, UseFormTrigger } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import useAuth from '../../../../hooks/useAuth';
import useErrorHandlers from '../../../../hooks/useErrorHandlers';
import signupSchema from '../validators/signupSchema';

interface UseSignupForm {
  loading: boolean;
  isValid: boolean;
  errors: Partial<FieldErrorsImpl<SignupFormInputs>>;
  control: Control<SignupFormInputs, unknown>;
  handleGoogleSignUp: () => void;
  submitForm: (e: React.FormEvent) => void;
  trigger: UseFormTrigger<SignupFormInputs>;
}

export type SignupFormInputs = {
  email: string;
  password: string;
  confirmPassword: string;
};

export const useSignupForm = (): UseSignupForm => {
  const {
    control,
    handleSubmit,
    trigger,
    setError,
    formState: { errors, isValid },
  } = useForm<SignupFormInputs>({
    mode: 'onChange',
    resolver: yupResolver(signupSchema),
  });
  const [loading, setLoading] = useState(false);
  const { signUp } = useAuth();
  const { handleRequestError } = useErrorHandlers();
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<SignupFormInputs> = useCallback(
    async ({ email, password }) => {
      try {
        setLoading(true);
        await signUp({ email, password });
        navigate('/');
      } catch (err) {
        const requestErrors = handleRequestError(err as AxiosError);
        if (requestErrors) {
          setError(
            'email',
            {
              message: requestErrors.message[0].constraints.isEmailExists,
            },
            {
              shouldFocus: true,
            }
          );
        }
      } finally {
        setLoading(false);
      }
    },
    [signUp, navigate, handleRequestError, setError]
  );

  const handleGoogleSignUp = useCallback(() => {
    window.location.href = `${import.meta.env.VITE_API_ENDPOINT}/auth/google`;
  }, []);

  const submitForm = (e: React.FormEvent) => {
    e.preventDefault();
    handleSubmit(onSubmit)();
  };

  return {
    control,
    errors,
    handleGoogleSignUp,
    isValid,
    loading,
    submitForm,
    trigger,
  };
};
