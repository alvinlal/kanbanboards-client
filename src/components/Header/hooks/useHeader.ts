import { useAuth } from '../../../hooks/useAuth/useAuth';
import { useToggleLeftSideNav } from '../../../hooks/useToggleLeftSideNav/useToggleLeftSideNav';

interface UseHeader {
  toggleLeftSideNav: () => void;
  logOut: () => Promise<void>;
}

export const useHeader = (): UseHeader => {
  const { logOut } = useAuth();
  const { toggleLeftSideNav } = useToggleLeftSideNav();

  return {
    toggleLeftSideNav,
    logOut,
  };
};
