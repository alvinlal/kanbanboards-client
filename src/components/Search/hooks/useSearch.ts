import throttle from 'lodash.throttle';
import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import search from '../../../api/search';
import SearchResponse from '../../../api/search/types/SearchResponse';

interface UseSearch {
  results: SearchResponse | null;
  isResultsVisible: boolean;
  setIsResultsVisible: React.Dispatch<React.SetStateAction<boolean>>;
  loading: boolean;
  searchInputRef: React.RefObject<HTMLInputElement>;
  throttledHandleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const useSearch = (): UseSearch => {
  const [results, setResults] = useState<SearchResponse | null>(null);
  const [isResultsVisible, setIsResultsVisible] = useState(false);
  const [loading, setLoading] = useState(false);
  const searchInputRef = useRef<HTMLInputElement>(null);
  const abortController = useMemo(() => new AbortController(), []);

  const handleChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      if (!e.target.value) {
        setResults(null);
        return;
      }
      try {
        setLoading(true);
        const { data } = await search(e.target.value, abortController.signal);
        if (searchInputRef.current?.value) {
          setResults(data);
        }
      } catch (err) {
        // send error to monitoring service
      } finally {
        setLoading(false);
      }
    },
    [abortController.signal]
  );

  const throttledHandleChange = useMemo(
    () => (e: React.ChangeEvent<HTMLInputElement>) => {
      const throttledFn = throttle(handleChange, 300);
      throttledFn(e);
    },
    [handleChange]
  );

  useEffect(() => {
    return () => {
      abortController.abort();
    };
  }, [abortController]);

  return {
    results,
    isResultsVisible,
    setIsResultsVisible,
    loading,
    searchInputRef,
    throttledHandleChange,
  };
};
