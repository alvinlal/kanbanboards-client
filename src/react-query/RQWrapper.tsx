/* eslint-disable no-console */
// this wrapper is for testing only

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import React from 'react';

const queryClient = new QueryClient({
  logger: {
    log: console.log,
    warn: console.warn,
    error: () => {},
  },
});
export const RQWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
