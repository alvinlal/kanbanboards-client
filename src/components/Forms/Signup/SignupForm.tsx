import { Controller } from 'react-hook-form';

import Button from '../../Button/Button';
import TextField from '../../Inputs/TextField/TextField';
import styles from './SignupForm.module.scss';
import { ReactComponent as GoogleLogo } from '../../../assets/icons/google.svg';

import Spinner from '../../Spinner/Spinner';
import { useSignupForm } from './hooks/useSignupForm';

type SignupFormProps = React.FormHTMLAttributes<HTMLFormElement>;

const SignupForm: React.FC<SignupFormProps> = (props) => {
  const {
    errors,
    isValid,
    control,
    handleGoogleSignUp,
    handleSubmit,
    onSubmit,
    loading,
    trigger,
  } = useSignupForm();

  return (
    <form
      className={styles.signup__form}
      onSubmit={handleSubmit(onSubmit)}
      {...props}
    >
      <h1>Signup</h1>
      <Controller
        control={control}
        name="email"
        render={({ field: { ref, onChange, onBlur } }) => (
          <TextField
            type="text"
            placeholder="email"
            error={!!errors.email}
            onChange={onChange}
            onBlur={onBlur}
            ref={ref}
          />
        )}
      />
      <div className={styles.error}>
        {errors.email && <p>{errors.email.message}</p>}
      </div>
      <Controller
        control={control}
        name="password"
        render={({ field: { ref, onChange, onBlur } }) => (
          <TextField
            type="password"
            placeholder="password"
            error={!!errors.password}
            onChange={async (e) => {
              onChange(e);
              await trigger('confirmPassword');
            }}
            onBlur={onBlur}
            ref={ref}
          />
        )}
      />
      <div className={styles.error}>
        {errors.password && <p>{errors.password.message}</p>}
      </div>
      <Controller
        control={control}
        name="confirmPassword"
        render={({ field: { ref, onChange, onBlur } }) => (
          <TextField
            type="password"
            placeholder="confirm password"
            error={!!errors.confirmPassword}
            onChange={onChange}
            onBlur={onBlur}
            ref={ref}
          />
        )}
      />
      <div className={styles.error}>
        {errors.confirmPassword && <p>{errors.confirmPassword.message}</p>}
      </div>
      <Button
        type="submit"
        width="100%"
        height="56px"
        variant="primary"
        disabled={!isValid || loading}
      >
        <p>SIGNUP</p>
        {loading && <Spinner />}
      </Button>
      <hr />
      <Button
        type="button"
        onClick={handleGoogleSignUp}
        width="100%"
        height="56px"
        variant="secondary"
      >
        <p>CONTINUE WITH GOOGLE</p>
        <GoogleLogo width={28} height={28} />
      </Button>
    </form>
  );
};

export default SignupForm;
