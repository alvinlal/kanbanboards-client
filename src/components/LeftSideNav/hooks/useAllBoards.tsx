import { useQuery } from '@tanstack/react-query';
import { getAllBoards } from '../../../api/board';
import { AllBoardsResponseDto } from '../../../api/board/dto/response/AllBoardsResponse.dto';
import queryKeys from '../../../react-query/queryKeys';

interface UseAllBoards {
  allBoards: AllBoardsResponseDto | undefined;
  isLoading: boolean;
  isError: boolean;
}

export const useAllBoards = (): UseAllBoards => {
  const { data: allBoards, isLoading, isError } = useQuery([queryKeys.allBoards], getAllBoards);

  return {
    allBoards,
    isLoading,
    isError,
  };
};
