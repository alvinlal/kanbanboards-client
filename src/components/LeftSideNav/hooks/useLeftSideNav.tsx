import { useMemo } from 'react';
import { AllBoardsResponse } from '../../../api/board/dto/response/AllBoardsResponse.dto';
import { useAllBoards } from '../../../hooks/useAllBoards/useAllBoards';
import { useAuth } from '../../../hooks/useAuth/useAuth';
import { useToggleLeftSideNav } from '../../../hooks/useToggleLeftSideNav/useToggleLeftSideNav';
import SkeletonBlock from '../../SkeletonBlock/SkeletonBlock';

interface UseLeftSideNav {
  LoadingBlocks: React.ReactNode[];
  allBoards: AllBoardsResponse | undefined;
  isLoading: boolean;
  isError: boolean;
  isLeftSideNavOpen: boolean;
  toggleLeftSideNav: () => void;
  logOut: () => Promise<void>;
}

export const useLeftSideNav = (): UseLeftSideNav => {
  const { allBoards, isLoading, isError } = useAllBoards();
  const { isLeftSideNavOpen, toggleLeftSideNav } = useToggleLeftSideNav();
  const { logOut } = useAuth();

  const LoadingBlocks = useMemo(() => {
    const blocks = [];

    for (let i = 0; i < 6; i += 1) {
      blocks.push(<SkeletonBlock width="200px" height="30px" key={i} />);
    }

    return blocks;
  }, []);

  return {
    LoadingBlocks,
    allBoards,
    isLoading,
    isError,
    isLeftSideNavOpen,
    toggleLeftSideNav,
    logOut,
  };
};
