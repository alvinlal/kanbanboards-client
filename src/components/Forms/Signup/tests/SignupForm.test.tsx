/* eslint-disable @typescript-eslint/ban-ts-comment */
/**
 * Requirements :-
 * Should show error on invalid email
 * Should show error on invalid password
 * Should show error if passwords don't match
 * Should re-check if passwords match on changing password field after filling confirm password field
 * Should show error if email already exists on clicking signup
 * Signup button should be disabled if there are errors
 * Should show loading spinner on clicking signup
 * Should not show any errors if all fields are valid
 * Should navigate user to home page after successfull signup
 * Should match snapshot
 * Should match loading snapshot
 * Should match error snapshot
 */
import { faker } from '@faker-js/faker';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';
import { apiEndPoint } from '../../../../../test-utils/msw/baseUrls';
import { signupController } from '../../../../../test-utils/msw/handlers/auth/controllers/signupController';
import { rest, server } from '../../../../../test-utils/msw/server';
import { RQWrapper } from '../../../../react-query/RQWrapper';
import { SignupFormInputs } from '../hooks/useSignupForm';
import SignupForm from '../SignupForm';
import defaultHandlers from '../../../../../test-utils/msw/defaultHandlers';

describe('SignupForm.tsx', () => {
  const typeIntoForm = async ({ email, password, confirmPassword }: Partial<SignupFormInputs>) => {
    const user = userEvent.setup();
    const emailInputElement = screen.getByLabelText('email') as HTMLInputElement;
    const passwordInputElement = screen.getByLabelText('password') as HTMLInputElement;
    const confirmPasswordInputElement = screen.getByLabelText(
      'confirm password'
    ) as HTMLInputElement;

    if (email) {
      await user.type(emailInputElement, email);
    }
    if (password) {
      await user.type(passwordInputElement, password);
    }
    if (confirmPassword) {
      await user.type(confirmPasswordInputElement, confirmPassword);
    }

    const emailInvalidErrorElement = screen.queryByText(/email is not valid/i);
    const passwordLengthErrorElement = screen.queryByText(
      /password should contain atleast 6 characters/i
    );
    const passwordSpaceErrorElement = screen.queryByText(/password should not contain spaces/i);
    const confirmPasswordDontMatchErrorElement = screen.queryByText(/passwords must match/i);

    return {
      emailInputElement,
      passwordInputElement,
      confirmPasswordInputElement,
      emailInvalidErrorElement,
      passwordLengthErrorElement,
      passwordSpaceErrorElement,
      confirmPasswordDontMatchErrorElement,
    };
  };

  const clickOnSignupButton = async () => {
    const signupButtonElement = screen.getByRole('button', {
      name: /SIGNUP/i,
    }) as HTMLButtonElement;
    await userEvent.click(signupButtonElement);
    return signupButtonElement;
  };

  const clickOnGoogleButton = async () => {
    const googleButtonElement = screen.getByRole('button', {
      name: /CONTINUE WITH GOOGLE/i,
    }) as HTMLButtonElement;
    await userEvent.click(googleButtonElement);
    return googleButtonElement;
  };

  describe('Should show email errors', () => {
    describe('when email is empty', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { emailInputElement } = await typeIntoForm({ email: faker.internet.email() });
        await userEvent.clear(emailInputElement);
        const emailIsRequiredErrorElement = await screen.findByText(/email is required/i);
        expect(emailIsRequiredErrorElement).toBeInTheDocument();
      });
    });

    describe('when email is invalid', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { emailInvalidErrorElement } = await typeIntoForm({ email: 'invalidemail.com' });
        expect(emailInvalidErrorElement).toBeInTheDocument();
      });
    });

    describe('when email already exists', () => {
      it('shows correct error message', async () => {
        server.use(rest.post(apiEndPoint('/auth/signup'), signupController[400]));

        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(6);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
        await clickOnSignupButton();
        const emailExistsErrorElement = await screen.findByText(
          /Account already exists, please login/i
        );
        expect(emailExistsErrorElement).toBeInTheDocument();
      });
    });
  });

  describe('Should show password errors', () => {
    describe('when password is empty', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { passwordInputElement } = await typeIntoForm({
          password: faker.internet.password(),
        });
        await userEvent.clear(passwordInputElement);
        const passwordIsRequiredErrorElement = await screen.findByText(/password is required/i);
        expect(passwordIsRequiredErrorElement).toBeInTheDocument();
      });
    });
    describe('when password is less than 6 characters', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { passwordLengthErrorElement } = await typeIntoForm({
          password: faker.internet.password(3),
        });
        expect(passwordLengthErrorElement).toBeInTheDocument();
      });
    });

    describe('when password contain space', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { passwordSpaceErrorElement } = await typeIntoForm({
          password: `${faker.internet.password(3)} `,
        });
        expect(passwordSpaceErrorElement).toBeInTheDocument();
      });
    });
  });

  describe('Should show confirm password errors', () => {
    describe('when confirm password is empty', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { confirmPasswordInputElement } = await typeIntoForm({
          confirmPassword: faker.internet.password(),
        });
        await userEvent.clear(confirmPasswordInputElement);
        const confirmPasswordIsRequiredErrorElement = await screen.findByText(
          /please confirm password/i
        );
        expect(confirmPasswordIsRequiredErrorElement).toBeInTheDocument();
      });
    });

    describe("when passwords don't match", () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(6);
        const { confirmPasswordDontMatchErrorElement } = await typeIntoForm({
          password,
          confirmPassword: `${password}123`,
        });
        expect(confirmPasswordDontMatchErrorElement).toBeInTheDocument();
      });
    });
  });

  describe('Should navigate user to the correct destination after successfull signup', () => {
    describe('when signup button is clicked with correct form inputs', () => {
      it('should navigate user to home page', async () => {
        const history = createMemoryHistory();
        history.push = jest.fn();

        render(
          <Router location={history.location} navigator={history}>
            <SignupForm />
          </Router>,
          { wrapper: RQWrapper }
        );

        const password = faker.internet.password(6);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });

        await clickOnSignupButton();

        expect(history.push).toHaveBeenCalledWith(
          { pathname: '/', search: '', hash: '' },
          undefined,
          {}
        );
      });

      it('should show loading spinner in signup button', async () => {
        server.use(rest.post(apiEndPoint('/auth/signup'), defaultHandlers.LOADING));
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(6);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
        await clickOnSignupButton();
        const spinner = await screen.findByTestId('spinner');
        expect(spinner).toBeInTheDocument();
      });

      it('signup button should be disabled while signup api request is loading', async () => {
        server.use(rest.post(apiEndPoint('/auth/signup'), defaultHandlers.LOADING));
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(6);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
        const signupButtonElement = await clickOnSignupButton();
        await waitFor(() => {
          expect(signupButtonElement.disabled).toBe(true);
        });
      });
    });

    describe('when continue with google button is clicked', () => {
      it('should navigate user to correct url', async () => {
        const assignMock = jest.fn();

        // @ts-ignore
        delete window.location;
        // @ts-ignore
        window.location = { assign: assignMock };

        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );

        await clickOnGoogleButton();
        expect(window.location.href).toBe(`${import.meta.env.VITE_API_ENDPOINT}/auth/google`);
      });
    });
  });

  describe('Should show correct button states for signup button', () => {
    describe('when there is no errors', () => {
      it('signup button should not be disabled', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(6);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
        const signupButtonElement = await clickOnSignupButton();
        await waitFor(() => {
          expect(signupButtonElement.disabled).toBe(false);
        });
      });
    });

    describe('when there are errors', () => {
      it('signup button should be disabled', async () => {
        render(
          <MemoryRouter>
            <SignupForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const password = faker.internet.password(2);
        await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
        const signupButtonElement = await clickOnSignupButton();
        await waitFor(() => {
          expect(signupButtonElement.disabled).toBe(true);
        });
      });
    });
  });

  describe('Should match snapshots', () => {
    it("match's default snapshot", () => {
      const tree = renderer.create(
        <MemoryRouter>
          <RQWrapper>
            <SignupForm />
          </RQWrapper>
        </MemoryRouter>
      );
      expect(tree).toMatchSnapshot();
    });
    it("match's loading state snapshot", async () => {
      server.use(rest.post(apiEndPoint('/auth/signup'), defaultHandlers.LOADING));

      const { asFragment } = render(
        <MemoryRouter>
          <SignupForm />
        </MemoryRouter>,
        {
          wrapper: RQWrapper,
        }
      );

      const password = faker.internet.password(6);
      await typeIntoForm({ email: faker.internet.email(), password, confirmPassword: password });
      await clickOnSignupButton();

      await waitFor(async () => {
        const spinner = await screen.findByTestId('spinner');
        expect(spinner).toBeInTheDocument();
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it("match's field error state snapshot", async () => {
      const { asFragment } = render(
        <MemoryRouter>
          <SignupForm />
        </MemoryRouter>,
        {
          wrapper: RQWrapper,
        }
      );
      const password = faker.internet.password(2);
      await typeIntoForm({
        email: 'invalidemail.com',
        password,
        confirmPassword: `${password}123`,
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
