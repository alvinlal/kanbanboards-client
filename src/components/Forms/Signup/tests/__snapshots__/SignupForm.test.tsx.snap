// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm.tsx Should match snapshots match's default snapshot 1`] = `
<form
  className="signup__form"
  onSubmit={[Function]}
>
  <h1>
    Signup
  </h1>
  <input
    aria-label="email"
    autoFocus={true}
    className="textfield null"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="email"
    type="text"
  />
  <div
    className="error"
  />
  <input
    aria-label="password"
    className="textfield null"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="password"
    type="password"
  />
  <div
    className="error"
  />
  <input
    aria-label="confirm password"
    className="textfield null"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="confirm password"
    type="password"
  />
  <div
    className="error"
  />
  <button
    className="button primary "
    disabled={true}
    style={
      {
        "height": "56px",
        "width": "100%",
      }
    }
    type="submit"
  >
    <p>
      SIGNUP
    </p>
  </button>
  <hr />
  <button
    className="button secondary "
    onClick={[Function]}
    style={
      {
        "height": "56px",
        "width": "100%",
      }
    }
    type="button"
  >
    <p>
      CONTINUE WITH GOOGLE
    </p>
    <div
      height={28}
      width={28}
    />
  </button>
</form>
`;

exports[`SignupForm.tsx Should match snapshots match's field error state snapshot 1`] = `
<DocumentFragment>
  <form
    class="signup__form"
  >
    <h1>
      Signup
    </h1>
    <input
      aria-label="email"
      class="textfield error"
      placeholder="email"
      type="text"
    />
    <div
      class="error"
    >
      <p>
        email is not valid
      </p>
    </div>
    <input
      aria-label="password"
      class="textfield error"
      placeholder="password"
      type="password"
    />
    <div
      class="error"
    >
      <p>
        password should contain atleast 6 characters
      </p>
    </div>
    <input
      aria-label="confirm password"
      class="textfield error"
      placeholder="confirm password"
      type="password"
    />
    <div
      class="error"
    >
      <p>
        passwords must match
      </p>
    </div>
    <button
      class="button primary "
      disabled=""
      style="width: 100%; height: 56px;"
      type="submit"
    >
      <p>
        SIGNUP
      </p>
    </button>
    <hr />
    <button
      class="button secondary "
      style="width: 100%; height: 56px;"
      type="button"
    >
      <p>
        CONTINUE WITH GOOGLE
      </p>
      <div
        height="28"
        width="28"
      />
    </button>
  </form>
</DocumentFragment>
`;

exports[`SignupForm.tsx Should match snapshots match's loading state snapshot 1`] = `
<DocumentFragment>
  <form
    class="signup__form"
  >
    <h1>
      Signup
    </h1>
    <input
      aria-label="email"
      class="textfield null"
      placeholder="email"
      type="text"
    />
    <div
      class="error"
    />
    <input
      aria-label="password"
      class="textfield null"
      placeholder="password"
      type="password"
    />
    <div
      class="error"
    />
    <input
      aria-label="confirm password"
      class="textfield null"
      placeholder="confirm password"
      type="password"
    />
    <div
      class="error"
    />
    <button
      class="button primary "
      disabled=""
      style="width: 100%; height: 56px;"
      type="submit"
    >
      <p>
        SIGNUP
      </p>
      <div
        class="spinner"
        data-testid="spinner"
      />
    </button>
    <hr />
    <button
      class="button secondary "
      style="width: 100%; height: 56px;"
      type="button"
    >
      <p>
        CONTINUE WITH GOOGLE
      </p>
      <div
        height="28"
        width="28"
      />
    </button>
  </form>
</DocumentFragment>
`;
